cpp/src/rolling/detail/range_utils.cuh

```diff
@@ -0,0 +1,869 @@
+ */
+template <typename T>
+struct less {
+  constexpr bool operator()(T const& x, T const& y) const noexcept
+  {
+    if constexpr (cuda::std::is_floating_point_v<T>) {
+      if (cuda::std::isnan(x)) { return false; }
+      return cuda::std::isnan(y) || x < y;
+    } else {
+      return x < y;
+    }
+  }
+};
+
+template <typename T>
+struct less_equal {
+  constexpr bool operator()(T const& x, T const& y) const noexcept
+  {
+    if constexpr (cuda::std::is_floating_point_v<T>) {
+      if (cuda::std::isnan(x)) { return cuda::std::isnan(y); }
+      return cuda::std::isnan(y) || x <= y;
+    } else {
+      return x <= y;
+    }
+  }
+};
+template <typename T>
+struct greater {
+  constexpr bool operator()(T const& x, T const& y) const noexcept
+  {
+    if constexpr (cuda::std::is_floating_point_v<T>) {
+      if (cuda::std::isnan(x)) { return !cuda::std::isnan(y); }
+      return !cuda::std::isnan(y) && x > y;
+    } else {
+      return x > y;
+    }
+  }
+};
+
+template <typename T>
+struct greater_equal {
+  constexpr bool operator()(T const& x, T const& y) const noexcept
+  {
+    if constexpr (cuda::std::is_floating_point_v<T>) {
+      if (cuda::std::isnan(x)) { return true; }
+      return !cuda::std::isnan(y) && x >= y;
+    } else {
+      return x >= y;
+    }
+  }
+};
+
+/**
+ * @brief Select the appropriate ordering comparator for the window type.
+ *
+ * @tparam T The type being compared.
+ * @tparam Tag The type of the window.
+ */
+template <typename T, window_tag Tag>
+struct comparator_impl {
+  using op     = void;
+  using rev_op = void;
+};
+
+template <typename T>
+struct comparator_impl<T, window_tag::BOUNDED_CLOSED> {
+  using op     = less<T>;
+  using rev_op = greater<T>;
+};
+template <typename T>
+struct comparator_impl<T, window_tag::BOUNDED_OPEN> {
+  using op     = less_equal<T>;
+  using rev_op = greater_equal<T>;
+};
+template <typename T>
+struct comparator_impl<T, window_tag::CURRENT_ROW> {
+  using op     = less<T>;
+  using rev_op = greater<T>;
+};
+
+/**
+ * @brief Select the appropriate ordering comparator for the window type.
+ *
+ * @tparam Tag The type of the window.
+ * @tparam Order The sort order of the column used to define the windows.
+ */
+template <typename T, window_tag Tag, order Order>
+using comparator_t = std::conditional_t<Order == order::ASCENDING,
+                                        typename comparator_impl<T, Tag>::op,
+                                        typename comparator_impl<T, Tag>::rev_op>;
+
+/**
+ * @brief Compute `x + y` saturating at the numeric bounds rather than
+ * overflowing.
+ *
+ * @tparam T the type of the result and left operand.
+ * @tparam V the type of the right operand.
+ * @param x The left operand.
+ * @param y The right operand.
+ *
+ * @returns Pair of `x + y`, saturated at the numeric limits for the type of
+ * `x`, without overflowing or invoking undefined behaviour, and
+ * whether overflow occurred.
```

Comment:
detecting the overflow is necessary for `bounded_open` windows where the actual window bound turns out to extend past the numeric limit for the type.

Reply:
Yes we do. The user is only allowed to pass a scalar offset, so there's no preprocessing they can do that would avoid an overflow here. Both pandas and polars support endpoint-excluded windows so we need a way to distinguish between "got the limit but overflowed" and "got the limit but didn't overflow" (as it happens, I think neither of them handle this correctly, but that's a bug I do not want to replicate)